facebookresearch/ParlAI
# ELI5, Download pushshift Reddit and CC Dumps

**Patch description**
Adding the ELI5 dataset and capability for downloading specific Reddit and Common Crawl dumps  as in #2094. There's an overview of changes for each ELI5 repo file in the other information section.
 
**Testing steps**
Testing `build.py`:
Run `python examples/display_data.py -t eli5` and checked directory structure and pre-computed files. `eli5` data folder had `eli5/pre_computed`, `eli5/processed_data`, `eli5/tmp`,
`eli5/processed_data/collected_docs` directories as expected. `wet.paths`, `explainlikeimfive_ccrawls_ids.json`, and `explainlikeimfive_unigram_counts` were also downloaded in `pre_computed` as expected.

Testing `agents.py`:
Took https://github.com/facebookresearch/ELI5/blob/master/data_creation/processed_data/public_examples/examples_qda_short.json since I can't run all the steps, and saved it as `eli5/processed_data/selected_15_1/explainlikeimfive_train.json`.
Ran `python examples/display_data.py -t eli5`,  single episode here:
```
6 . A HERBIVORE is a living thing that eats only plants . One example of this is a DEER . [... I removed a bunch of lines to make this smaller ...] Herbivores are also known as herbivorous animals .
If an animal is an herbivore ( only eats plants ) , where does it get the protein that it needs to survive if it only eats grass ? --T-- 
   Herbivores only eat plants , but unlike humans they have special digestive systems that let them get protein from grass . Plants contain cellulose , which is difficult for humans to digest . However , animals like cows have special digestive systems that can process the cellulose and then access nutrition like protein that are protected by the cellulose in plant cells . A lot of herbivores have multiple stomachs ( like cows ) or special areas in the stomach or intestine that can help in digesting plants .
```

Ran `python examples/display_data.py --knowledge False -t eli5`, single episode here:
```
If an animal is an herbivore ( only eats plants ) , where does it get the protein that it needs to survive if it only eats grass ? --T-- 
   Herbivores only eat plants , but unlike humans they have special digestive systems that let them get protein from grass . Plants contain cellulose , which is difficult for humans to digest . However , animals like cows have special digestive systems that can process the cellulose and then access nutrition like protein that are protected by the cellulose in plant cells . A lot of herbivores have multiple stomachs ( like cows ) or special areas in the stomach or intestine that can help in digesting plants .
```

To test downloading Reddit post ids:
`python3 download_reddit_qalist.py -id_l posts.json -Q`
`python3 download_reddit_qalist.py -id_l posts.json -A`
and checked the jsons in `processed_data/` had the posts and comments in them.
`posts.json` is a .json file containing the following:  `["fhm9mp", "fhpcke"]`
Also ran `python3 download_reddit_qalist.py -Q -sy 2011 -ey 2011` and `python3 download_reddit_qalist.py -A -sy 2011 -ey 2011` to make sure those still worked.

Testing `download_support_docs.py`:
Edited `wet.paths` to only contain the following:
`crawl-data/CC-MAIN-2018-34/segments/1534221208676.20/wet/CC-MAIN-20180814062251-20180814082251-00000.warc.wet.gz`

Ran `python3 parlai/tasks/eli5/download_support_docs.py -u urls.json`
where `urls.json` contained the text ` ["https://openeffect.com/"]`
and checked that the `slices/0/`'s 0000 slice contained the openeffect url, id, and text.

Ran `python3 parlai/tasks/eli5/download_support_docs.py -ids ids.json`
where `ids.json` contained the text ` ["aec9fe35-7fec-46b3-96de-dd2db8887902"]` or alternatively `["<urn:uuid:aec9fe35-7fec-46b3-96de-dd2db8887902>"]` (this is the CCID corresponding with openeffect.com)
and checked that the `slices/0/`'s 0000 slice contained the openeffect url, id, and text.
Slice content for both for reference:
```
[[0, {"ccid": "<urn:uuid:aec9fe35-7fec-46b3-96de-dd2db8887902>", "url": "http://openeffect.com/", "text": ["This site requires JavaScript and Cookies to be enabled . Please change your browser settings or upgrade your browser .", []]}]]
```
Testing `merge_support_docs.py`:
Ran `python3 parlai/tasks/eli5/merge_support_docs.py -c 0 -n specific_ids` and
`python3 parlai/tasks/eli5/merge_support_docs.py -c finalize -n specific_ids`
Checked `0.json` and `slices/slice_0.json` had same text as above slice.

**Other information**
Overview of Changes from github.com/facebookresearch/ELI5/:
`download_support_docs.py`: Changed to use ParlAI parser and datapath, added ability to download specific urls and ids from Common Crawl
Added `select_specific_ids`, `select_specific_urls`, `check_url`, code to differentiate between downloading specific posts, ccids, or eli5 docs. The `main` code for downloading and saving docs was from ELI5.

`download_reddit_qalist.py`: Changed to use ParlAI parser and datapath, added ability to download specific reddit post ids
Added `download_and_process_comments`, `download_and_process_posts`, `get_posts`, `get_comments_from_post`, and a few parts of the parsing part main for downloading specific reddit posts. Rest is from ELI5.

`merge_support_docs.py`: Changed to use ParlAI parser and datapath. Added `merge_non_eli_docs` and a few lines in `main` to merge non-eli5 docs.

`data_creation/data_utils.py`: No changes except lint
`data_creation/finalize_qda.py`,  `/select_sentences_tfidf.py`: Changed to use ParlAI parser and datapath
`data_creation/slurm_scripts/eli_download_docs.sh`, `/eli_merge_docs.sh`: Changed to use relative `data_creation` directory rather than hard-coded directory
`data_creation/slurm_scripts/eli_select_docs.sh`: Changed to use relative `data_creation` directory rather than hard-coded directory, changed to use parser arguments instead of sysv args (due to `select-sentences_tfidf.py` changes)
All other files in `slurm_scripts/` were unchanged
`README.md`: Changes to mention the ELI5 repo, mention automatic downloading of pre-computed files, section at end for general reddit downloading and common crawl ids/urls.

**Data tests (if applicable)**
If you added a new teacher, you will be asked to run
`python setup.py test -s tests.suites.datatests`. Please paste this log here.

I don't have the resources to download everything and run the slurm scripts, so I can't confirm the final dataset.
